processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java

```diff
@@ -0,0 +1,283 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.join;
+
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import org.apache.druid.java.util.common.granularity.Granularity;
+import org.apache.druid.java.util.common.guava.Sequence;
+import org.apache.druid.java.util.common.guava.Sequences;
+import org.apache.druid.query.QueryMetrics;
+import org.apache.druid.query.filter.Filter;
+import org.apache.druid.segment.Capabilities;
+import org.apache.druid.segment.Cursor;
+import org.apache.druid.segment.Metadata;
+import org.apache.druid.segment.StorageAdapter;
+import org.apache.druid.segment.VirtualColumn;
+import org.apache.druid.segment.VirtualColumns;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.apache.druid.segment.data.Indexed;
+import org.apache.druid.segment.data.ListIndexed;
+import org.joda.time.DateTime;
+import org.joda.time.Interval;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+public class HashJoinSegmentStorageAdapter implements StorageAdapter
+{
+  private final StorageAdapter baseAdapter;
+  private final List<JoinableClause> clauses;
+
+  HashJoinSegmentStorageAdapter(
+      StorageAdapter baseAdapter,
+      List<JoinableClause> clauses
+  )
+  {
+    this.baseAdapter = baseAdapter;
+    this.clauses = clauses;
+  }
+
+  @Override
+  public Interval getInterval()
+  {
+    return baseAdapter.getInterval();
+  }
+
+  @Override
+  public Indexed<String> getAvailableDimensions()
+  {
+    // Use a Set since we may encounter duplicates, if a field from a Joinable shadows one of the base fields.
+    final LinkedHashSet<String> availableDimensions = new LinkedHashSet<>();
+
+    baseAdapter.getAvailableDimensions().forEach(availableDimensions::add);
+
+    for (JoinableClause clause : clauses) {
+      clause.getJoinable()
+            .getAvailableColumns()
+            .stream()
+            .map(c -> clause.getPrefix() + c)
+            .forEach(availableDimensions::add);
```

Comment:
This probably should be a method on JoinableClause.

Reply:
esp given the `unPrefix` method which assumes the logic in the `map` call above.

Reply:
And you'd probably need a unit test to make sure that what the clause prefixes is unprefixed correctly.

Reply:
These changes sound like good ideas. I'll make them.